### PR TITLE
Feature/added events and metrics to scheme adapter

### PR DESCRIPTION
**Pull request by rmothilal**
Added support for events and prometheus metrics

### CODE REVIEW

File: src/InboundServer/api.yaml [hunk 1]
```diff
@@ -2643,6 +2738,26 @@ components:
         see API Signature document.
       schema:
         type: string
+    traceparent:
+      name: traceparent
+      in: header
+      required: false
+      description: >-
+        The traceparent header field is a non-HTTP standard field used by the API
```
Comment (mdebarros) [inline src/InboundServer/api.yaml]: Please refer to the following --> https://mojaloop.io/documentation/mojaloop-technical-overview/event-framework/#411-wc3-http-headers.

Also, please grab the definition from the official WC3 spec and provide a link to where people can read more information on the header.

File: src/InboundServer/api.yaml [hunk 2]
```diff
@@ -2643,6 +2738,26 @@ components:
...
+        for passing along the parent traceId for children traces to use to create child
+        spans.
+      schema:
+        type: string
+    tracestate:
+      name: tracestate
+      in: header
+      required: false
+      description: >-
+        The tracestate header field is a non-HTTP standard field used by the API
```
Comment (mdebarros) [inline src/InboundServer/api.yaml]: Please refer to the following --> https://mojaloop.io/documentation/mojaloop-technical-overview/event-framework/#411-wc3-http-headers

Also, please grab the definition from the official WC3 spec and provide a link to where people can read more information on the header.

File: src/InboundServer/api.yaml [hunk 3]
```diff
@@ -2643,6 +2738,26 @@ components:
...
+        for passing along the current state of the parent traceId being passed in
+        the traceparent header
+      schema:
+        type: string
```
Comment (mdebarros) [inline src/InboundServer/api.yaml]: See pattern/schema based on the WC3 standard.

File: src/InboundServer/api.yaml [hunk 4]
```diff
@@ -2643,6 +2738,26 @@ components:
...
+    traceparent:
+      name: traceparent
+      in: header
+      required: false
+      description: >-
+        The traceparent header field is a non-HTTP standard field used by the API
+        for passing along the parent traceId for children traces to use to create child
+        spans.
+      schema:
+        type: string
```
Comment (mdebarros) [inline src/InboundServer/api.yaml]: See pattern/schema based on the WC3 standard.

File: src/InboundServer/handlers.js [hunk 5]
```diff
@@ -351,21 +417,27 @@ const putPartiesByTypeAndIdError = async(ctx) => {
...
+    histTimerEnd({ success: true });
 };
 
 
 /**
  * Handles a PUT /quotes/{ID}/error request. This is an error response to a POST /quotes request
  */
 const putQuotesByIdError = async(ctx) => {
+    const histTimerEnd = Metrics.getHistogram(
+        'inbound_put_quotes_id_error',
```
Comment (mdebarros) [inline src/InboundServer/handlers.js]: Any reason that you are not using the information that you set in the Route Map instead....it seems that this is using the same metric names for the event framework?

File: src/InboundServer/handlers.js [hunk 6]
```diff
@@ -323,21 +383,27 @@ const putTransfersById = async (ctx) => {
...
+    histTimerEnd({ success: true });
 };
 
 
 /**
  * Handles a PUT /parties/{Type}/{ID}/error request. This is an error response to a GET /parties/{Type}/{ID} request
  */
 const putPartiesByTypeAndIdError = async(ctx) => {
+    const histTimerEnd = Metrics.getHistogram(
+        'inbound_put_parties_type_id_error',
```
Comment (mdebarros) [inline src/InboundServer/handlers.js]: Any reason that you are not using the information that you set in the Route Map instead....it seems that this is using the same metric names for the event framework?

File: src/InboundServer/handlers.js [hunk 7]
```diff
@@ -299,21 +353,27 @@ const putQuoteById = async (ctx) => {
...
+    histTimerEnd({ success: true });
 };
 
 
 /**
  * Handles a PUT /transfers/{ID}. This is a response to a POST /transfers request
  */
 const putTransfersById = async (ctx) => {
+    const histTimerEnd = Metrics.getHistogram(
+        'inbound_put_transfers_id',
```
Comment (mdebarros) [inline src/InboundServer/handlers.js]: Any reason that you are not using the information that you set in the Route Map instead....it seems that this is using the same metric names for the event framework?

File: src/InboundServer/handlers.js [hunk 8]
```diff
@@ -274,21 +322,27 @@ const putPartiesByTypeAndId = async (ctx) => {
...
+    histTimerEnd({ success: true });
 };
 
 
 /**
  * Handles a PUT /quotes/{ID}. This is a response to a POST /quotes request
  */
 const putQuoteById = async (ctx) => {
+    const histTimerEnd = Metrics.getHistogram(
+        'inbound_put_quotes_id',
```
Comment (mdebarros) [inline src/InboundServer/handlers.js]: Any reason that you are not using the information that you set in the Route Map instead....it seems that this is using the same metric names for the event framework?

File: src/InboundServer/handlers.js [hunk 9]
```diff
@@ -257,14 +300,19 @@ const putParticipantsByTypeAndId = async (ctx) => {
  * request.
  */
 const putPartiesByTypeAndId = async (ctx) => {
+    const histTimerEnd = Metrics.getHistogram(
+        'inbound_put_parties_type_id',
```
Comment (mdebarros) [inline src/InboundServer/handlers.js]: Any reason that you are not using the information that you set in the Route Map instead....it seems that this is using the same metric names for the event framework?

File: src/InboundServer/handlers.js [hunk 10]
```diff
@@ -214,13 +250,19 @@ const putParticipantsById = async (ctx) => {
...
+    histTimerEnd({ success: true });
 };
 
 
 /**
  * Handles a PUT /participants/{ID}/error. This is an error response to a POST /participants request
  */
 const putParticipantsByIdError = async (ctx) => {
+    const histTimerEnd = Metrics.getHistogram(
+        'inbound_put_participants_type_id_error',
```
Comment (mdebarros) [inline src/InboundServer/handlers.js]: Any reason that you are not using the information that you set in the Route Map instead....it seems that this is using the same metric names for the event framework?

File: src/InboundServer/handlers.js [hunk 11]
```diff
@@ -191,12 +221,18 @@ const postTransfers = async (ctx) => {
...
     ctx.response.body = '';
+    histTimerEnd({ success: true });
 };
 
 /**
  * Handles a PUT /participants/{ID}. This is a response to a POST /participants request
  */
 const putParticipantsById = async (ctx) => {
+    const histTimerEnd = Metrics.getHistogram(
+        'inbound_put_participants_id',
```
Comment (mdebarros) [inline src/InboundServer/handlers.js]: Any reason that you are not using the information that you set in the Route Map instead....it seems that this is using the same metric names for the event framework?

File: src/InboundServer/handlers.js [hunk 12]
```diff
@@ -146,13 +168,19 @@ const postQuotes = async (ctx) => {
...
+    histTimerEnd({ success: true });
 };
 
 
 /**
  * Handles a POST /transfers request
  */
 const postTransfers = async (ctx) => {
+    const histTimerEnd = Metrics.getHistogram(
+        'inbound_post_transfers',
```
Comment (mdebarros) [inline src/InboundServer/handlers.js]: Any reason that you are not using the information that you set in the Route Map instead....it seems that this is using the same metric names for the event framework?

File: src/InboundServer/handlers.js [hunk 13]
```diff
@@ -108,6 +124,11 @@ const postPartiesByTypeAndId = (ctx) => {
  * Handles a POST /quotes request
  */
 const postQuotes = async (ctx) => {
+    const histTimerEnd = Metrics.getHistogram(
+        'inbound_post_quotes',
```
Comment (mdebarros) [inline src/InboundServer/handlers.js]: Any reason that you are not using the information that you set in the Route Map instead....it seems that this is using the same metric names for the event framework?

File: src/InboundServer/handlers.js [hunk 14]
```diff
@@ -55,6 +63,11 @@ const getParticipantsByTypeAndId = async (ctx) => {
  */
 const getPartiesByTypeAndId = async (ctx) => {
     // kick off an asyncronous operation to handle the request
+    const histTimerEnd = Metrics.getHistogram(
+        'inbound_get_parties_type_id',
```
Comment (mdebarros) [inline src/InboundServer/handlers.js]: Any reason that you are not using the information that you set in the Route Map instead....it seems that this is using the same metric names for the event framework?

File: src/InboundServer/handlers.js [hunk 15]
```diff
@@ -13,11 +13,17 @@
...
 const Model = require('@internal/model').InboundTransfersModel;
 const { Errors } = require('@mojaloop/sdk-standard-components');
+const Metrics = require('@mojaloop/central-services-metrics');
 
 /**
  * Handles a GET /participants/{idType}/{idValue} request
  */
 const getParticipantsByTypeAndId = async (ctx) => {
+    const histTimerEnd = Metrics.getHistogram(
+        'inbound_get_participants_type_id',
```
Comment (mdebarros) [inline src/InboundServer/handlers.js]: Any reason that you are not using the information that you set in the Route Map instead....it seems that this is using the same metric names for the event framework?

File: src/InboundServer/handlers.js [hunk 16]
```diff
@@ -376,21 +448,27 @@ const putQuotesByIdError = async(ctx) => {
...
+    histTimerEnd({ success: true });
 };
 
 
 /**
  * Handles a PUT /transfers/{ID}/error. This is an error response to a POST /transfers request
  */
 const putTransfersByIdError = async (ctx) => {
+    const histTimerEnd = Metrics.getHistogram(
+        'inbound_put_transfers_id_error',
```
Comment (mdebarros) [inline src/InboundServer/handlers.js]: Any reason that you are not using the information that you set in the Route Map instead....it seems that this is using the same metric names for the event framework?

File: src/InboundServer/handlers.js [hunk 17]
```diff
@@ -455,52 +534,131 @@ const getCallbackById = async(ctx) => {
...
+            enableSpan: true
+        }
     },
     '/participants/{Type}/{ID}': {
-        put: putParticipantsByTypeAndId,
-        get: getParticipantsByTypeAndId
+        put: {
+            handler: putParticipantsByTypeAndId,
+            id: 'inbound_put_participants_type_id',
+            enableSpan: true
```
Comment (mdebarros) [inline src/InboundServer/handlers.js]: `enalbeSpan` -- should probably be --> `enableTracing` <--- tracing is the important part, we will create spans to enable the tracing.

Let me know what you think.

Same comment applies to the rest of the map entries.

File: src/OutboundServer/handlers.js [hunk 18]
```diff
@@ -47,6 +48,12 @@ const handleAccountsError = (method, err, ctx) =>
  * Handler for outbound transfer request initiation
  */
 const postTransfers = async (ctx) => {
+    const histTimerEnd = Metrics.getHistogram(
+        'outbound_post_transfers',
```
Comment (mdebarros) [inline src/OutboundServer/handlers.js]: Same comment here about the Route Maps having a duplicate metric name, and the rest of the entries below?

File: src/OutboundServer/handlers.js [hunk 19]
```diff
@@ -140,19 +164,47 @@ const healthCheck = async (ctx) => {
...
+};
+
 module.exports = {
     map: {
         '/': {
-            get: healthCheck
+            get: {
+                handler: healthCheck,
+                id: 'outbound_health_check',
+                enableSpan: false
```
Comment (mdebarros) [inline src/OutboundServer/handlers.js]: Same comment here applies to the previous InboundAPI comments.

File: src/config.js [hunk 20]
```diff
@@ -73,6 +73,19 @@ let DEFAULTS = {
...
+    metrics: {
+        disabled: true,
+        labels: {
+            fspId: '*'
+        },
+        config: {
+            timeout: 5000,
+            prefix: 'moja_sdk_',
+            defaultLabels: {
+                serviceName: 'mojaloop_sdk'
```
Comment (mdebarros) [inline src/config.js]: I think this should reflect the component. I.e. `sdk-schema-adapter`. What service names do we use for central-ledger and ml-api-adapter?

File: src/lib/middleware/middleware.js [hunk 21]
```diff
@@ -0,0 +1,63 @@
+/**************************************************************************
```
Comment (mdebarros) [inline src/lib/middleware/middleware.js]: Shouldn't this middleware have an appropriate name to indicate that it's for `tracing`?

File: src/lib/middleware/package.json [hunk 22]
```diff
@@ -0,0 +1,21 @@
+{
```
Comment (mdebarros) [inline src/lib/middleware/package.json]: Shouldn't this middleware have an appropriate name to indicate that it's for `tracing`?

File: src/lib/model/InboundTransfersModel.js [hunk 23]
```diff
@@ -65,24 +66,34 @@ class InboundTransfersModel {
      * Queries the backend API for the specified party and makes a callback to the originator with our dfspId if found
      */
     async getParticipantsByTypeAndId(idType, idValue, sourceFspId) {
+        const histTimerEnd = Metrics.getHistogram(
+            'inbound_get_participants_type_id',
```
Comment (mdebarros) [inline src/lib/model/InboundTransfersModel.js]: Same comment applies here for metric re-use based on map, and for the rest of the file.

File: src/lib/model/OutboundTransfersModel.js [hunk 24]
```diff
@@ -166,15 +170,22 @@ class OutboundTransfersModel {
      * then waits for a notification from the cache that the payee has been resolved.
      */
     async _resolvePayee() {
+        const histTimerEnd = Metrics.getHistogram(
+            'outbound_resolve_payee',
```
Comment (mdebarros) [inline src/lib/model/OutboundTransfersModel.js]: Same comment applies here for metric re-use based on map and the rest of the file.

File: src/config.js [hunk 25]
```diff
@@ -205,7 +218,22 @@ const setConfig = async cfg => {
 
     config.logIndent = cfg.LOG_INDENT ? Number(cfg.LOG_INDENT) : 2;
 
+    config.metrics = {
+        disabled: cfg.METRICS_DISABLED.toLowerCase() === 'true',
```
Comment (mdebarros) [inline src/config.js]: I assume these are being defaulted as per the section above if they are not set by the ENV?

File: src/local.env [hunk 26]
```diff
@@ -118,3 +118,11 @@ ALS_ENDPOINT=127.0.0.1:6500
...
 # To allow transfer without a previous quote request, set this value to true and provide a matching fulfilment for the condition
 ALLOW_TRANSFER_WITHOUT_QUOTE=false
+
+# Configuration for the metrics library to acquire metrics for prometheus
+# Is disabled by default
+METRICS_DISABLED=false
+METRICS_LABELS_FSP_ID=*
+METRICS_CONFIG_TIMEOUT=5000
+METRICS_CONFIG_PREFIX=moja_sdk_
+METRICS_CONFIG_LABELS_SERVICE_NAME=mojaloop-sdk
```
Comment (mdebarros) [inline src/local.env]: Same comment here applies as per the default configuration.

File: src/package.json [hunk 27]
```diff
@@ -19,13 +19,17 @@
   "dependencies": {
     "@internal/cache": "file:lib/cache",
     "@internal/log": "file:lib/log",
+    "@internal/middleware": "file:lib/middleware",
```
Comment (mdebarros) [inline src/package.json]: I really think this should be something specific as `middleware` can mean anything. I.e. why isn't this `tracing-middleware`, unless you think the same component will be used to define other middleware components?

File: src/InboundServer/handlers.js (hunk 5 above)
Comment (rmothilal) [reply src/InboundServer/handlers.js]: decided not to go with this option as name in route map requires the path to access

File: src/lib/model/InboundTransfersModel.js (hunk 23 above)
Comment (rmothilal) [reply src/lib/model/InboundTransfersModel.js]: decided not to go with this option as name in route map requires the path to access

File: src/lib/model/OutboundTransfersModel.js (hunk 24 above)
Comment (rmothilal) [reply src/lib/model/OutboundTransfersModel.js]: decided not to go with this option as name in route map requires the path to access

File: src/config.js (hunk 25 above)
Comment (rmothilal) [reply src/config.js]: Yes it is

File: src/InboundServer/api.yaml (hunk 4 above)
Comment (rmothilal) [reply src/InboundServer/api.yaml]: pattern added

File: src/InboundServer/api.yaml (hunk 3 above)
Comment (rmothilal) [reply src/InboundServer/api.yaml]: pattern added

File: src/InboundServer/api.yaml (hunk 1 above)
Comment (rmothilal) [reply src/InboundServer/api.yaml]: updated

File: src/InboundServer/handlers.js (hunk 6 above)
Comment (rmothilal) [reply src/InboundServer/handlers.js]: decided not to go with this option as name in route map requires the path to access

File: src/InboundServer/handlers.js (hunk 8 above)
Comment (rmothilal) [reply src/InboundServer/handlers.js]: decided not to go with this option as name in route map requires the path to access

File: src/InboundServer/handlers.js (hunk 7 above)
Comment (rmothilal) [reply src/InboundServer/handlers.js]: decided not to go with this option as name in route map requires the path to access

File: src/InboundServer/handlers.js (hunk 9 above)
Comment (rmothilal) [reply src/InboundServer/handlers.js]: decided not to go with this option as name in route map requires the path to access

File: src/InboundServer/handlers.js (hunk 10 above)
Comment (rmothilal) [reply src/InboundServer/handlers.js]: decided not to go with this option as name in route map requires the path to access

File: src/InboundServer/handlers.js (hunk 11 above)
Comment (rmothilal) [reply src/InboundServer/handlers.js]: decided not to go with this option as name in route map requires the path to access

File: src/InboundServer/handlers.js (hunk 12 above)
Comment (rmothilal) [reply src/InboundServer/handlers.js]: decided not to go with this option as name in route map requires the path to access

File: src/InboundServer/handlers.js (hunk 14 above)
Comment (rmothilal) [reply src/InboundServer/handlers.js]: decided not to go with this option as name in route map requires the path to access

File: src/InboundServer/handlers.js (hunk 13 above)
Comment (rmothilal) [reply src/InboundServer/handlers.js]: decided not to go with this option as name in route map requires the path to access

File: src/InboundServer/handlers.js (hunk 15 above)
Comment (rmothilal) [reply src/InboundServer/handlers.js]: decided not to go with this option as name in route map requires the path to access

File: src/InboundServer/handlers.js (hunk 16 above)
Comment (rmothilal) [reply src/InboundServer/handlers.js]: decided not to go with this option as name in route map requires the path to access

File: src/OutboundServer/handlers.js (hunk 18 above)
Comment (rmothilal) [reply src/OutboundServer/handlers.js]: decided not to go with this option as name in route map requires the path to access

File: src/OutboundServer/handlers.js (hunk 19 above)
Comment (rmothilal) [reply src/OutboundServer/handlers.js]: decided not to go with this option as name in route map requires the path to access

File: src/InboundServer/handlers.js (hunk 17 above)
Comment (rmothilal) [reply src/InboundServer/handlers.js]: updated

**Pull request merged**